icsm-au/DynAdjust
# Add .travis build

I had a go at prototyping a way to run automated builds & tests using continuous integration. Let me know what you think.

I've used a service called Travis to do this (which is free for open source projects). There are other services like it, but most large open source software projects seem to use Travis (for example [QGIS](https://github.com/qgis/QGIS/blob/master/.travis.yml)).

Travis is essentially an online virtual machine which hooks into GitHub. So now, every time a new commit is made to GitHub

1. Travis downloads the latest code
2. looks for a [`.travis` file](https://github.com/jmettes/DynAdjust/blob/ce10bcceed5b1699791b896ccabad2d9323c17d1/.travis.yml)
3. executes its instructions

I've set it up to build and test on every commit. To see what this looks like on Travis, check out https://travis-ci.org/jmettes/DynAdjust/builds/524874865.

---

Steps followed in the [`.travis` file](https://github.com/jmettes/DynAdjust/blob/ce10bcceed5b1699791b896ccabad2d9323c17d1/.travis.yml):

1. install system packages for the build dependencies
    - operating system packages change a lot, this might be a way to quickly find out when things break

2. build DynAdjust
    - if at any stage it fails to build, the whole pipeline is stopped
    - this gives a strong incentive to fix the bug

3. perform very crude system test
    - it compares the output of an adjustment to a previously run adjustment, given identical inputs
    -  this shouldn't be different, unless either
        1. a change in the commit broke the software (see [example](https://travis-ci.org/jmettes/DynAdjust/jobs/524868050#L1594) of me breaking it)
        2. or the software was changed enough to warrant an update to the test
    - it would be ideal to have more granular unit tests to drill down to error sources. but for now, I'd say this is better than nothing

This whole process takes about 15 minutes. This might become too long in the future. I'm playing around with some ideas to shave off some of this time.

---

Further notes:

- Builds on Ubuntu. I believe Travis also supports Windows builds
    - not sure yet how to go about testing on NCI
- I've added a small subset of AUSGeoid2020_20170829 (bounding box of stations in sample data)
